### ISSUE TITLE
datetime.fromisoformat() only accepts 3 or 6 decimal places for fractional seconds

### ISSUE THREAD

**Issue by rdfguy** [python/cpython#95221]
The ISO 8601 spec allows any number of decimal places for fractional seconds, but the `datetime.fromisoformat()` method raises a `ValueError` if there is a fractional part that has something other than 3 or 6 decimal places (I checked the source code). This causes problems with, for example, Adobe XMP metadata.

This raises a `ValueError`:
```
import datetime
datetime.datetime.fromisoformat("2021-06-26T08:34:41.12-05:00")  # only two decimal places
```

Tested in this environment:

```
Python 3.8.3 (v3.8.3:6f8c8320e9, May 13 2020, 16:29:34) 
[Clang 6.0 (clang-600.0.57)] on darwin
```

**Comment by tiran** [comment]
``2021-06-26T08:34:41.12-05:00`` is an ISO format string with a time zone (``-05:00``, probably eastern standard time).

**Comment by rdfguy** [comment]
Except that it *does* support time zones. The only real limitation I have found, namely that only 3 or 6 decimal places for the fractional second is allowed, is *rather arbitrary* and causes the function to fail when parsing ISO date stamps produced by other systems. There is no good justification for this, and it would be easy to fix.

**Comment by markaidev** [comment]
I verified the bug exists in old versions of python, but when I forked, compiled, and made it from the newest source, the errors are gone. I am new to this codebase so maybe I missed something, but it seems to be fixed somehow.

**Comment by abalkin** [comment]
As @tiran noted [above](https://github.com/python/cpython/issues/95221#issuecomment-1193902407), the limitation was intentional when the fromisoformat method was introduced in 3.7 and well-documented. See GH-60077. The supported format prior to 3.11 was

```
YYYY-MM-DD[*HH[:MM[:SS[.fff[fff]]]][+HH:MM[:SS[.ffffff]]]]
```

See [datetime.fromisoformat](https://docs.python.org/3.10/library/datetime.html?highlight=fromisoformat#datetime.datetime.fromisoformat). Since the requested behaviour is already implemented in 3.11, I am closing this issue.
